web/src/scheduled_messages.js

```diff
@@ -0,0 +1,197 @@
+    popover_menus.show_schedule_confirm_button(scheduled_msg.formatted_send_at_time, true);
+}
+
+function format(scheduled_messages) {
+    const formatted_msgs = [];
+    for (const msg of scheduled_messages) {
+        const msg_render_context = {...msg};
+        if (msg.type === "stream") {
+            msg_render_context.is_stream = true;
+            msg_render_context.stream_id = msg.to[0];
+            msg_render_context.stream_name = stream_data.maybe_get_stream_name(
+                msg_render_context.stream_id,
+            );
+            const color = stream_data.get_color(msg_render_context.stream_name);
+            msg_render_context.recipient_bar_color = stream_color.get_recipient_bar_color(color);
+            msg_render_context.stream_privacy_icon_color =
+                stream_color.get_stream_privacy_icon_color(color);
+        } else {
+            msg_render_context.is_stream = false;
+            msg_render_context.recipients = people.get_recipients(msg.to.join(","));
+        }
+        const time = new Date(msg.deliver_at);
+        msg_render_context.full_date_time = timerender.get_full_datetime(time);
+        msg_render_context.formatted_send_at_time = date_fns.format(time, "MMM d yyyy h:mm a");
+        formatted_msgs.push(msg_render_context);
+    }
+    return formatted_msgs;
+}
+
+export function delete_scheduled_message(scheduled_msg_id) {
+    channel.del({
+        url: "/json/scheduled_messages/" + scheduled_msg_id,
+        success() {
+            // TODO: Do this via events received from the server in server_events_dispatch.
+            if (overlays.scheduled_messages_open()) {
+                $(
+                    `#scheduled_messages_overlay .scheduled-message-row[data-message-id=${scheduled_msg_id}]`,
+                ).remove();
+            }
+            if ($("#compose-textarea").attr("data-scheduled-message-id")) {
+                const compose_scheduled_msg_id = $("#compose-textarea").attr(
+                    "data-scheduled-message-id",
+                );
+                // If user deleted the scheduled message which is being edited in compose, we clear
+                // the scheduled message id from there which converts this editing state into a normal
+                // schedule message state. So, clicking "Schedule" will now create a new scheduled message.
+                if (compose_scheduled_msg_id === scheduled_msg_id) {
+                    $("#compose-textarea").removeAttr("data-scheduled-message-id");
+                }
+            }
+        },
+    });
+}
+
+export function delete_scheduled_message_if_sent_directly() {
+    // Delete old scheduled message if it was sent.
+    if ($("#compose-textarea").attr("data-scheduled-message-id")) {
+        delete_scheduled_message($("#compose-textarea").attr("data-scheduled-message-id"));
+        $("#compose-textarea").removeAttr("data-scheduled-message-id");
+    }
+}
```

Comment:
If it's easy, it might be worth splitting the code here between two files; a `scheduled_messages_overlay_ui.js` and a `scheduled_messages.js`; the latter would just have the API code like `delete_scheduled_message` and functions like this one, but as we add things like events handling would grow to have all the data structures logic (and be organized with `reminder.js` in some sane way), with this `_overlay_ui.js` module being reserved for the overlay's UI logic.

(We'll almost certainly need to do that sort of split for long term import cycle reasons eventually, and it might be nice to just do that while we're working on this now, since it looks like 80% of this module is overlay stuff anyway).

Reply:
Done.